Pinta.Core/Managers/ToolManager.cs

```diff
@@ -172,23 +185,32 @@ public void SetCurrentTool (Gdk.Key shortcut)
 			if (tool != null)
 				SetCurrentTool(tool);
 		}
-		
+
 		private BaseTool FindNextTool (Gdk.Key shortcut)
 		{
-			string key = shortcut.ToString ().ToUpperInvariant ();
-			
-			// Begin looking at the tool after the current one
-			for (int i = index + 1; i < Tools.Count; i++) {
-				if (Tools[i].ShortcutKey.ToString ().ToUpperInvariant () == key)
-					return Tools[i];
-			}
-				
-			// Begin at the beginning and look up to the current tool
-			for (int i = 0; i < index; i++) {
-				if (Tools[i].ShortcutKey.ToString ().ToUpperInvariant () == key)
-					return Tools[i];
+			foreach (var group in groupedTools)
+			{
+				if (group.Key.EqualsTo(shortcut))
+				{
+					//begin looking for the tool from the start
+					if (group.Value.Item2 >= group.Value.Item1.Count)
```

Comment:
Could this extra section be avoided if the code below handled wrapping the index? (i.e. updating to `(i + 1) % count` instead of just `++i`). I think the loop could also be avoided if the index is always valid.

Reply:
Yes, you are right. Made it as you say.